### PR TITLE
As the user I want to see details of an application of a candidate so that I can check the candidate history

### PR DESCRIPTION
**Pull request by kbertalan**

### CODE REVIEW

File: persistence/src/main/resources/db/migration/data/integration/V1_12_1__AddStatesToStateTable.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+INSERT INTO states (id,name) VALUES (1,'newstate');
```
Comment (kbertalan) [inline persistence/src/main/resources/db/migration/data/integration/V1_12_1__AddStatesToStateTable.sql]: This SQL migration file contains mission critical data, without it Atsy would not work.

In this sense this SQL migration should be placed into the schema folder in order to every environment (production and integration) would be able to built upon it.

File: persistence/src/main/resources/db/migration/data/integration/V1_12_2__AddStateFlow.sql [hunk 2]
```diff
@@ -0,0 +1,64 @@
+INSERT INTO state_flow (from_id,to_id) VALUES (1,2);
```
Comment (kbertalan) [inline persistence/src/main/resources/db/migration/data/integration/V1_12_2__AddStateFlow.sql]: Like for the state values, this SQL migration script should be placed into the schema folder.

File: persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StatesRepositoryIT.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.persistence.repositories;
+/*package com.epam.rft.atsy.persistence.repositories;
```
Comment (kbertalan) [inline persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StatesRepositoryIT.java]: Please maintain the existing test code.
New repositories should have tests as well.
The pull request will not be accepted without tests.

File: service/src/main/java/com/epam/rft/atsy/service/domain/states/AbstractStateDTO.java [hunk 4]
```diff
@@ -1,6 +1,7 @@
 package com.epam.rft.atsy.service.domain.states;
```
Comment (kbertalan) [inline service/src/main/java/com/epam/rft/atsy/service/domain/states/AbstractStateDTO.java:1]: I think this DTO should follow the rename of the State -> StateHistory, so it should be AbstractStateHistoryDTO

With the rename the oddity of having a StateDTO in an AbstractStateDTO would vanish.

File: service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java [hunk 5]
```diff
@@ -71,35 +78,40 @@
...
     ApplicationEntity applicationEntity = applicationsRepository.findOne(applicationId);
 
     Assert.notNull(applicationEntity);
-    List<StateEntity>
+    List<StatesHistoryEntity>
         stateEntities =
-        statesRepository.findByApplicationEntityOrderByStateIndexDesc(applicationEntity);
-    List<StateViewDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
+        statesHistoryRepository.findByApplicationEntityOrderByCreationDateDesc(applicationEntity);
+    List<StateViewHistoryDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
```
Comment (kbertalan) [inline service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java]: Also rename the variable.

File: service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java [hunk 6]
```diff
@@ -71,35 +78,40 @@
...
 
   @Override
-  public List<StateViewDTO> getStatesByApplicationId(Long applicationId) {
+  public List<StateViewHistoryDTO> getStatesByApplicationId(Long applicationId) {
     Assert.notNull(applicationId);
     ApplicationEntity applicationEntity = applicationsRepository.findOne(applicationId);
 
     Assert.notNull(applicationEntity);
-    List<StateEntity>
+    List<StatesHistoryEntity>
```
Comment (kbertalan) [inline service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java]: Also rename the variable.

File: service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java [hunk 7]
```diff
@@ -71,35 +78,40 @@
...
         stateEntities =
-        statesRepository.findByApplicationEntityOrderByStateIndexDesc(applicationEntity);
-    List<StateViewDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
+        statesHistoryRepository.findByApplicationEntityOrderByCreationDateDesc(applicationEntity);
+    List<StateViewHistoryDTO> stateDTOs = modelMapper.map(stateEntities, STATEVIEWDTO_LIST_TYPE);
 
     SimpleDateFormat simpleDateFormat = new SimpleDateFormat(DATE_FORMAT_CONSTANT);
 
     for (int i = 0; i < stateDTOs.size(); i++) {
```
Comment (kbertalan) [inline service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java]: As we don't know which implementation is used for the List interface here, the get(i) operations can have performance penalty. Switch to foreach instead.

File: service/src/test/java/com/epam/rft/atsy/service/impl/ApplicationsServiceImplTest.java [hunk 8]
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.service.impl;
+/*package com.epam.rft.atsy.service.impl;
```
Comment (kbertalan) [inline service/src/test/java/com/epam/rft/atsy/service/impl/ApplicationsServiceImplTest.java]: Maintain the existing tests.

File: service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java [hunk 9]
```diff
@@ -1,4 +1,4 @@
-package com.epam.rft.atsy.service.impl;
+/*package com.epam.rft.atsy.service.impl;
```
Comment (kbertalan) [inline service/src/test/java/com/epam/rft/atsy/service/impl/StatesServiceImplTest.java]: Maintain the existing tests.

File: web/src/main/java/com/epam/rft/atsy/web/controllers/ApplicationStateController.java [hunk 10]
```diff
@@ -1,21 +1,49 @@
...
-  public ModelAndView loadPage(@RequestParam Long applicationId) {
+  public ModelAndView loadPage(@RequestParam Long applicationId, Locale locale) {
     ModelAndView modelAndView = new ModelAndView(VIEW_NAME);
     modelAndView.addObject("applicationId", applicationId);
+
+    Collection<StateViewHistoryDTO>
+        applicationStates =
+        statesService.getStatesByApplicationId(applicationId);
+
+    for (StateViewHistoryDTO stateViewHistoryDTO : applicationStates) {
```
Comment (kbertalan) [inline web/src/main/java/com/epam/rft/atsy/web/controllers/ApplicationStateController.java]: This is the point where it is better to have a new class in the web layer which contains the state history view from the ui perspective.

You can convert the StateHistoryViewDTO instances into the StateHistoryViewXY object and put in the necessary messages.

File: service/src/main/java/com/epam/rft/atsy/service/domain/states/StateViewHistoryDTO.java [hunk 11]
```diff
@@ -14,7 +14,7 @@
 @AllArgsConstructor
 @EqualsAndHashCode(callSuper = true)
 @ToString(callSuper = true)
-public class StateViewDTO extends AbstractStateDTO {
+public class StateViewHistoryDTO extends AbstractStateDTO {
```
Comment (kbertalan) [inline service/src/main/java/com/epam/rft/atsy/service/domain/states/StateViewHistoryDTO.java]: Rename to StateHistoryViewDTO

File: web/src/main/webapp/WEB-INF/pages/application_state.jsp [hunk 12]
```diff
@@ -6,26 +6,246 @@
...
+      <div class="button-panel clearfix">
+          <form class="form-inline pull-right">
+              <c:if test="${states[0].stateDTO.name != reject && states[0].stateDTO.name != accept}">
+              <div class="form-group">
+                  <label class="form-label">Következő lépés:</label>
+              </div>
+              <div class="form-group">
+                  <div class="btn-group" role="group">
+                      <c:choose>
+                          <c:when test="${states[0].stateDTO.name == newstate}">
```
Comment (kbertalan) [inline web/src/main/webapp/WEB-INF/pages/application_state.jsp]: Don't use the localized message to find out which section should be displayed. Use the internal name of the state instead.

File: web/src/main/webapp/WEB-INF/pages/application_state.jsp [hunk 13]
```diff
@@ -6,26 +6,246 @@
...
+      <spring:message code="candidate.table.state.wageOffer" var="wageOffer" />
+      <spring:message code="candidate.table.state.agree" var="agree" />
+      <spring:message code="candidate.table.state.accept" var="accept" />
+      <spring:message code="candidate.table.state.reject" var="reject" />
+      <spring:message code="candidate.table.state.pause" var="pause" />
+      <div class="button-panel clearfix">
+          <form class="form-inline pull-right">
+              <c:if test="${states[0].stateDTO.name != reject && states[0].stateDTO.name != accept}">
+              <div class="form-group">
+                  <label class="form-label">Következő lépés:</label>
```
Comment (kbertalan) [inline web/src/main/webapp/WEB-INF/pages/application_state.jsp]: Don't leave hardcoded messages in the jsp-s

File: web/src/main/webapp/WEB-INF/pages/application_state.jsp [hunk 14]
```diff
@@ -6,26 +6,246 @@
...
+              </div>
+              <div class="form-group">
+                  <div class="btn-group" role="group">
+                      <c:choose>
+                          <c:when test="${states[0].stateDTO.name == newstate}">
+                              <a href="#" class="btn btn-default"><spring:message code="statehistory.buttons.cv"/></a>
+                              <a href="#" class="btn btn-warning"><spring:message code="statehistory.buttons.onHold"/></a>
+                              <a href="#" class="btn btn-danger"><spring:message code="statehistory.buttons.reject"/></a>
+                          </c:when>
+                          <c:when test="${states[0].stateDTO.name == cv}">
```
Comment (kbertalan) [inline web/src/main/webapp/WEB-INF/pages/application_state.jsp]: I thought the StateFlow entity and database table was introduced to be able to provide these sections dynamically.

**Comment by kbertalan** [comment]
In addition to the code comments, please add an SQL script which populates one application with state history records, so we could easily check them without the need to manually enter test data into database (while we don't have the proper functionality to create / modify application states)

File: persistence/src/main/resources/db/migration/data/integration/V1_12_1__AddStatesToStateTable.sql (hunk 1 above)
Comment (szabob94) [inline persistence/src/main/resources/db/migration/data/integration/V1_12_1__AddStatesToStateTable.sql]: fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/3731abffdcebc27e4fb7c9efdaf80d6c45fba90a

File: persistence/src/main/resources/db/migration/data/integration/V1_12_2__AddStateFlow.sql (hunk 2 above)
Comment (szabob94) [inline persistence/src/main/resources/db/migration/data/integration/V1_12_2__AddStateFlow.sql]: Fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/fe06da0fcae25b73c5f624b733c182f14e75ba48

File: service/src/main/java/com/epam/rft/atsy/service/domain/states/AbstractStateDTO.java (hunk 4 above)
Comment (szabob94) [inline service/src/main/java/com/epam/rft/atsy/service/domain/states/AbstractStateDTO.java:1]: fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

File: service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java (hunk 5 above)
Comment (szabob94) [inline service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java]: fixed by commit https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

File: service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java (hunk 6 above)
Comment (szabob94) [inline service/src/main/java/com/epam/rft/atsy/service/impl/StatesServiceImpl.java]: fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

File: service/src/main/java/com/epam/rft/atsy/service/domain/states/StateViewHistoryDTO.java (hunk 11 above)
Comment (szabob94) [inline service/src/main/java/com/epam/rft/atsy/service/domain/states/StateViewHistoryDTO.java]: fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/d9a0c08a56c02eb4db35ebec41d6e58d3dbf1834

File: service/src/test/java/com/epam/rft/atsy/service/impl/ApplicationsServiceImplTest.java (hunk 8 above)
Comment (szabob94) [inline service/src/test/java/com/epam/rft/atsy/service/impl/ApplicationsServiceImplTest.java]: fixed by https://github.com/epam-debrecen-rft-2015/atsy/commit/1f506e2848edeb4fb1ccfe72e2b1906875047d18

File: persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/ApplicationsRepositoryIT.java [hunk 15]
```diff
@@ -86,24 +88,38 @@ public void findByCandidateEntityShouldFindThreeApplicationForCandidateWithThree
...
         .name("Fejlesztő")
         .build();
     Date nearNow = currentDateMinus(5);
 
+    SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    Date expectedDate = null;
+    try {
+      expectedDate = simpleDateFormat.parse("2016-07-26 11:48:55");
+    } catch (ParseException e) {
+      e.printStackTrace();
```
Comment (kbertalan) [inline persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/ApplicationsRepositoryIT.java]: Never use printStackTrace.

In test usually you don't need to handle exceptions, if an exception occurs, than your test case will fail with error.

File: persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StateFlowRepositoryIT.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+          .toStateEntity(TO_STATES_ENTITY_PAUSE).build();
+  @Autowired
+  private StateFlowRepository stateFlowRepository;
+  @Autowired
+  private StatesRepository statesRepository;
+
+  @Test
+  public void findByFromStateEntityShouldNotFindStateFlowEntityListWhenInvalidFromStateEntityIsGiven() {
+    //Given
+    StatesEntity statesEntity = this.statesRepository.findOne(INVALID_STATE_ID);
```
Comment (kbertalan) [inline persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/StateFlowRepositoryIT.java]: As I remember this findOne returns null, so why don't you use statesEntity = null instead?

**Pull request merged**